### ISSUE TITLE
Set up JDK-13 build and drop jdk-12

### ISSUE THREAD

**Issue by pbludov** [checkstyle/checkstyle#7281]
Travis and appveyor already support Java 13, so we have to set up a build with it.

This is a prerequisite for #7103.

**Comment by romani** [comment]
OFF topic by we need to come to agreeement:
@rnveach , @pbludov , @strkkk , what is your thoughts on upgrading checkstyle to new minimal jdk version from 8 to 11 ? When it is better to do this ?

As jdk releases become a bit more predictable ... we can use some time trigger for example: upgrade to LTS after second minor release of LTS.
very conservative: stay on previous available LTS (ones 14th available, we upgrade to 11).

Should we use only LTS ? probably yes, as we are massively used in all sorts of projects. If we stay the most compatible, we will have more users and potential contributors.
Opposite side of such approach is that our project become less interesting for user who like modern features.

**Comment by rnveach** [comment]
I am against upgrading minimal JDK version. As you know I already maintain backport of checkstyle to Java 6. Even though I don't need as far back as Java 6 anymore, I will probably continue to use that.

**Comment by romani** [comment]
Please share reasons of not to upgrade ?
Your support of jdk6 port, is problem for you even now and future jdk upgrades will not make it very complicated, level of complexity will be the same, as even right now you have to rewrite code (lambda, streams).

**Comment by rnveach** [comment]
My reasons haven't changed since when we did the upgrade to 8. It will force people to upgrade for just a simple utility and those who can't or don't want to upgrade are forced to stay on an old version. I feel a utility should be supported by any jre and not only select versions.

**Comment by strkkk** [comment]
> what is your thoughts on upgrading checkstyle to new minimal jdk version from 8 to 11 ? When it is better to do this ?

In my opinion, not in the nearest feature. It will affect some users to upgrade its jdk, they can be not ready for it yet. 
There is already a backport for java6, so there is gonna be one more for java8.
Plus, I do not see much benefit (modules? var?) except that it is modern and cool.
To attract new users and contributors to the project some conference talks would be better.

> I feel a utility should be supported by any jre and not only select versions

I agree with this, some popular libraries (e.g https://github.com/stleary/JSON-java) still compiled with java 6

I think the right moment will be when java 11 will make it to 60%+ and drop java 6 support at the same time.
![image](https://user-images.githubusercontent.com/8901354/69849647-622d1880-128e-11ea-8556-9acdcb155963.png)
[source](https://www.jetbrains.com/lp/devecosystem-2019/java/)

**Comment by pbludov** [comment]
After a year, Oracle will stop releasing updates for Java 8. For many companies, this will become a security issue. I suppose we need to keep Java 8 as the main SDK until 2021. Then start the transition process to Java 11 and complete it before the release of Java 17 LTS.

**Comment by strkkk** [comment]
> Oracle will stop releasing updates for Java 8. For many companies, this will become a security issue

Obviously Oracle is the main, but not the only one vendor. For example, Amazon is gonna support java8 until 2023 at least.
Version 8 will be with us for long time :)

>Amazon will distribute security updates to Corretto 8 at no cost until at least June, 2023

https://aws.amazon.com/blogs/opensource/amazon-corretto-no-cost-distribution-openjdk-long-term-support/

**Comment by romani** [comment]
One more point of a lot jdk support, is CI time, we need to keep a lot of builds for different jdk

**Comment by romani** [comment]
upgrade to jdk8  - https://checkstyle.org/releasenotes.html#Release_7.0
in scope of https://github.com/checkstyle/checkstyle/issues/3229 - no that much details on why I did upgrade, probably based on some chart conversations, first jdk8 features were not used immediately ... our code still old styled.
https://checkstyle.org/releasenotes.html#Release_6.2 - switch to jdk7
https://checkstyle.org/releasenotes.html#Release_6.0 - switch to jdk6

original motivations on jdk update is at https://groups.google.com/forum/#!searchin/checkstyle-devel/jdk$20version%7Csort:date/checkstyle-devel/nd8rWCtJ7Tg/YU6sHPy5ydEJ

the most valuable reason for me is keep project development modern and interesting.
Do not fall in ancinet legacy support and keep coding on most old java to make some project happy
to use on the same jdk version.
There will be bunch of syntax improvments and jdk API extensions for more convenient coding.

If I stay as much as compatible as possible, I would never upgrade jdk.
checkstyle logic can work on jdk5, it is still probably possible to download it from somewhere and some projects still use jdk5.

----

Lets make experiement ..... imagine I did not do upgrade ... and checkstyle now on jdk5, all around the world enjoy this incredible effort of compatibility.
@pbludov and @strkkk, would you enjoy contribution to such project?
reminder about your desire use Optional and streams .... that is fancy style and not very critical ... but still good ... so why to limit ourself ? to make sombedoby happy to use checkstyle latest version. By helping other ... donot forget about yourself. 

-------

please keep in mind that legacy projects can still use latest checkstyle by execution of checkstyle 
in separate build on modern jdk(we are not runtime dependency, we are build time dependency), 
just for checkstyle ... if they care. IDE is runs on jdk1X but target jdk is jdk6 is completely ok model of development. We are not blocking them !!!!
we just make it a bit inconvenient, if they really want use latest, most users use some old version and are fine with this.

checkstyle maintanance is not a slavery, we do a lot to make development convenient and pleasant for us !!!

---

after recent changes in jdk (after 8), almost all migrated to openjdk (not oracle jdk) if possible, and it is become standard and open possibilities to keep own openjdk build, so openjdk11 will be long lasting  for download and use on your own risk.
openjdk version will awailable for LONG time, security issues are not critical for some projects
as there are a lot of proejcts who do not care about security at ALL 
as they are always run in secure environment (non public).
So we should not stay on jdk version that minor percent of users use.

Most users will use LTS version only, so I am completely fine to do the same.
Queston is only when to cut old versions

My proposal:
stay on previous to latest LTS always (then jdk11 released, stay on jdk 8, then jdk14 release migrate to jdk11)
so we will stay on not so old, but not an ancient. So actual projects will start migration to jdk14 from jdk11,
and we in this time migrate to jdk11. If some project stuck in old jdk version, they should not keep us.
Keep in mind migration from jdk8 to jdk11 will not that quick :) for us, we use so much tools so there will be bunch of nuances and blockers (coverage, pitest, plugins , ......). 
For example update of checkstyle maven plugin to use jdk8 parent pom ... took a while.
We do not have to immediately migrate to new JDK when trigger time happened ... but will not be blocked by maintainers ... and still migration should not sacrifice existing quality level. So we might run into blocker that will naturally/technologically block upgrade.

Please propose exact model on when to upgdate.... based on amazon end of life +- some period, based on IDEA statistics :) if it going to be updated (there is no 2018 statistcs), ... . 
Not to upgrade is not good strategy. 
On demand is also not good strategy as it will cause kind of the same discussion if somebody would like to use new switch syntax or instanceof syntax or ..... .

**Comment by pbludov** [comment]
@romani Java 14 won't be LTS. Next LTS release will be Java 17, scheduled at September 2021:
See https://en.wikipedia.org/wiki/Java_version_history

**Comment by romani** [comment]
LTS based periods still make sense for me, so checkstyle will be always 6-12 versions behind .... uhhh it is for sure more than enough. Are you agree ? 

We can start with such strategy and change it if find something very useful in recent versions and accelerate upgrade by special case.
I wish we can use more recent versions, but I am not sure what is good version to jump between LTS, but life will show us.

**Comment by pbludov** [comment]
I would like to use the new features from the new JDK, but this will break the build to anyone who has not yet switched to JDK 11+. This means that we need to fork and support one more version. I suppose this will have to be done anyway, but we need to wait for most users to switch from JDK 8.

**Comment by romani** [comment]
>  This means that we need to fork and support one more version. 

we do not need to do this. It is problem of user. 

> but we need to wait for most users to switch from JDK 8

please provide exact trigger.

**Comment by strkkk** [comment]
> we do not need to do this. It is problem of user.

Yeah, and I imagine the most common solution - do not update checkstyle version or do not use it. And it will cause need for backport for bugfixes. 
Tools like checkstyle should be more conservative and satisfy needs of most of the projects.

> please provide exact trigger.

Java 17 release, usage of java 8 in projects less than X% (by some statistics)

> Lets make experiement ..... imagine I did not do upgrade ... and checkstyle now on jdk5, all around the world enjoy this incredible effort of compatibility.
@pbludov and @strkkk, would you enjoy contribution to such project?

Interesting tasks are more attractive than some new features. Java 11 doesn't have features/things that are really worth upgrade and all the problems related to it.
Java 5 is extreme example. I am not saying there is no need at all, but it should be after most of the user projects migrate to not cause issues for them.

**Comment by pbludov** [comment]
@romani @strkkk @rnveach 
We can also drop JDK-9 and JDK-10 builds to speed up the CI process. For javac9 specific phase we can set up JDK-11 build with `--release 9`

Please vote.

**Comment by strkkk** [comment]
@pbludov I think it is ok. 
We can in future also exclude versions that are not supported anymore.

**Comment by rnveach** [comment]
> We can also drop JDK-9 and JDK-10 builds to speed up the CI process. 

IMO, a quick build and test of Checkstyle isn't that long a process, compared to Pitest execution and parallel runs of travis. I don't see dropping 1 or 2 items speeding anything up. If we want to drop them, then thats fine, but I don't see it hurting keeping them. For the most part I walk away when CI is running and don't wait for it minute by minute.

**Comment by rnveach** [comment]
Please confirm if this finalizes issue.

**Comment by romani** [comment]
@rnveach , according to https://travis-ci.org/checkstyle/checkstyle/builds/619800310 , custom jdk build takes 10 minutes for each (we have 2 build items). 
https://travis-ci.org/checkstyle/checkstyle/jobs/619800348
https://travis-ci.org/checkstyle/checkstyle/jobs/619800349
Do you really think we need such builds for jdk 9,10,12,13,14,15,16, ...  ?
we should presume that java is compatible at least till moment we know it is not in some cases :) .
so we keep only LTS - 8 and 11, for minor version only the latest.

**Comment by romani** [comment]
issue is fixed .... if any decision appear later on - new issue should be created.

**Comment by rnveach** [comment]
> https://travis-ci.org/checkstyle/checkstyle/jobs/619800348
https://travis-ci.org/checkstyle/checkstyle/jobs/619800349

Which contain downloads which up the time.

https://travis-ci.org/checkstyle/checkstyle/builds/618473073
https://travis-ci.org/checkstyle/checkstyle/jobs/618473118

Some runs look to be 4-5 minutes. Probably depends on how much is needed to be downloaded.

> Do you really think we need such builds for jdk 9,10,12,13,14,15,16, ... ?

I never stated we should add a run for each JDK. I just said removing 1 job won't instantly make CI run faster. We added 12 for a reason, so I thought we should keep it to make sure that reason doesn't resurface.

**Comment by romani** [comment]
We had such builds before for each jdk but I do not see a reason to have all of them, latest and lts are enough.

**Comment by romani** [comment]
in continuation of minimal JDK ....

we migrated to jdk8 at https://github.com/checkstyle/checkstyle/releases/tag/checkstyle-7.0 (Jun 26, 2016)

https://wiki.eclipse.org/Eclipse/Installation - jdk8 is minimal requirement from June 22, 2016. 
https://www.jetbrains.com/help/idea/installation-guide.html - jdk11 is minimal requirement.

:) awesome coincident :) almost in the same time as Eclipse. 
Most users use IDEA https://programmer9.com/2019/02/06/largest-java-survey-2018.html#which-ide-java-developers-use , but Eclipse usage is still big. What if we agree to migrate after Eclipse migrate to new jdk ?

**Comment by romani** [comment]
`SonarScanner will require Java 11 to run starting in SonarQube 8.x`
7.9 is current LTS now

from:
```
✔ ~/java/github/checkstyle/sonar-checkstyle [master|✚ 3] 
$ mvn sonar:sonar
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------< com.github.checkstyle:checkstyle-sonar-plugin >------------
[INFO] Building Checkstyle SonarQube Plugin 4.28-SNAPSHOT
[INFO] ----------------------------[ sonar-plugin ]----------------------------
[INFO] 
[INFO] --- sonar-maven-plugin:3.4.1.1168:sonar (default-cli) @ checkstyle-sonar-plugin ---
[INFO] User cache: /home/rivanov/.sonar/cache
[INFO] SonarQube version: 7.9.2
[INFO] Default locale: "en_US", source code encoding: "UTF-8"
[WARNING] SonarScanner will require Java 11 to run starting in SonarQube 8.x

```

**Comment by rnveach** [comment]
@pbludov Has there been any testing with JDK 13 and loading custom modules via reflection.

My work has a newer version JDK, 13, and I was testing it out for my work project and all my custom modules for checkstyle are broken and are saying it can't be found. I haven't dived into the issue more to find where the problem lies.

**Comment by romani** [comment]
https://sonarcloud.io/documentation/upcoming/ , at October 2020, sonar will stop accepting jdk8 executions

**Comment by pbludov** [comment]
So it's time to move on. We need this build to be successful:
```
 mvn -Djava.version=11 \
         -Dmaven.compiler.showWarnings=true \
         -Dmaven.compiler.showDeprecation=true \
         clean verify 
```
I'll start a new issue.
